lib/ai/model/tool.go

```diff
@@ -109,18 +134,281 @@ func (*commandExecutionTool) parseInput(input string) (*commandExecutionToolInpu
+			requestable[requestableRole] = struct{}{}
+		}
+	}
+	for _, role := range roles {
+		for _, requestableRole := range role.GetAccessRequestConditions(types.Deny).Roles {
+			delete(requestable, requestableRole)
+		}
+	}
+
+	resp := strings.Builder{}
+	for role := range requestable {
+		resp.Write([]byte(role))
+		resp.Write([]byte("\n"))
+	}
+
+	if resp.Len() == 0 {
+		return "No requestable roles found", nil
+	}
+
+	return resp.String(), nil
+}
+
+type accessRequestListRequestableResourcesTool struct{}
+
+func (*accessRequestListRequestableResourcesTool) Name() string {
+	return "List Requestable Resources"
 }
 
-type embeddingRetrievalTool struct {
-	assistClient      assist.AssistEmbeddingServiceClient
-	nodeClient        NodeGetter
-	userAccessChecker services.AccessChecker
-	currentUser       string
+func (*accessRequestListRequestableResourcesTool) Description() string {
+	return `List all resources with IDs that can be requested via access requests.
+This includes nodes via SSH access.`
 }
 
+func (a *accessRequestListRequestableResourcesTool) Run(ctx context.Context, toolCtx *ToolContext, input string) (string, error) {
+	foundResources := make(chan promptResource, 0)
+	workersAlive := new(int32)
+	*workersAlive = 5
+
+	searchAndAppend := func(resourceType string, convert func(types.Resource) (promptResource, error)) error {
+		defer func() {
+			if atomic.AddInt32(workersAlive, -1) == 0 {
+				close(foundResources)
+			}
+		}()
+
+		list, err := toolCtx.ListResources(ctx, proto.ListResourcesRequest{
+			ResourceType:     resourceType,
+			Limit:            maxShownRequestableItems,
+			UseSearchAsRoles: true,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		for _, resource := range list.Resources {
+			resource, err := convert(resource)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			select {
+			case foundResources <- resource:
+			case <-ctx.Done():
+				return trace.Wrap(ctx.Err())
+			}
+		}
+
+		return nil
+	}
+	searchAndAppendPlain := func(resourceType string) error {
+		return searchAndAppend(resourceType, func(resource types.Resource) (promptResource, error) {
+			return promptResource{
+				Name:    resource.GetName(),
+				Kind:    resource.GetKind(),
+				SubKind: resource.GetSubKind(),
+				Labels:  resource.GetMetadata().Labels,
+			}, nil
+		})
+	}
+
+	go searchAndAppend(types.KindNode, func(resource types.Resource) (promptResource, error) {
+		return promptResource{
+			Name:         resource.GetName(),
+			Kind:         resource.GetKind(),
+			SubKind:      resource.GetSubKind(),
+			Labels:       resource.GetMetadata().Labels,
+			FriendlyName: resource.(types.Server).GetHostname(),
+		}, nil
+	})
+	go searchAndAppendPlain(types.KindApp)
+	go searchAndAppendPlain(types.KindKubernetesCluster)
+	go searchAndAppendPlain(types.KindDatabase)
+	go searchAndAppendPlain(types.KindWindowsDesktop)
+
+	sb := strings.Builder{}
+	total := 0
+	for resource := range foundResources {
```

Comment:
Probably can be replaced with `for i < 5`

Reply:
We can't, we need to rely on the channel to notify here I believe. Otherwise you've got a logical race that can result in a deadlock.